_source/_posts/2022-02-23-angular-async-config.md

```diff
@@ -0,0 +1,348 @@
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
+
+import { AppModule } from './app/app.module';
+import { environment } from './environments/environment';
+import { OKTA_CONFIG } from '@okta/okta-angular';
+import { OktaAuth } from '@okta/okta-auth-js';
+
+if (environment.production) {
+  enableProdMode();
+}
+
+fetch('http://{yourApiUri}/config').then(async res => {
+  const authConfig = await res.json();
+
+  platformBrowserDynamic([
+    { provide: OKTA_CONFIG, useValue: {oktaAuth: new OktaAuth(authConfig)}}
+  ]).bootstrapModule(AppModule)
+    .catch(err => console.error(err));
+});
+```
+
+Then your `AppModule` only needs to import `OktaAuthModule` since you've already provided the `OKTA_CONFIG` injection token.
+
+If you need to create the callback URI programmatically or if you need to use the configuration in multiple places, you can store the configuration in the app instead. The minimum we need is a class that holds the config, which we will show in the example. You can wrap the config in a service if your needs are more involved than what we'll show here.
+
+You'll add a new file and create an interface that matches the response, as well as a class to hold the config:
+
+```ts
+export interface AuthConfig {
+  issuer: string;
+  clientId: string;
+}
+
+export class OktaAuthConfig {
+  constructor(public config: AuthConfig) { }
+}
+```
+
+Edit the `src/main.ts` to provide the `OktaAuthConfig` class instead
+
+```ts
+import { OktaAuthConfig } from './app/auth-config';
+
+fetch('http://{yourApiUri}/config').then(async res => {
+  const authConfig = new OktaAuthConfig(await res.json());
+
+  platformBrowserDynamic([
+    { provide: OktaAuthConfig, useValue: authConfig }
+  ]).bootstrapModule(AppModule)
+  .catch(err => console.error(err));
+})
+```
+
+In the `AppModule` you can provide the `OKTA_CONFIG` needed to integrate with Okta by accessing `OktaAuthConfig`:
+
+```ts
+@NgModule({
+  declarations: [ AppComponent, ProfileComponent ],
+  imports: [
+    BrowserModule,
+    AppRoutingModule,
+    OktaAuthModule
+  ],
+  providers: [
+    {
+      provide: OKTA_CONFIG,
+      deps: [OktaAuthConfig],
+      useFactory: (oktaConfig: OktaAuthConfig) => ({
+        oktaAuth: new OktaAuth({
+          ...oktaConfig.config,
+          redirectUri: window.location.origin + '/login/callback'
+        })
+      })
+    }
+  ]
+})
+export class AppModule { }
+```
+
+You can now load a config from an API and use the app's location.
+
+You might be asking yourself, "Isn't there an `APP_INITIALIZER` token or something we can use instead"? Well, yes, there is an [`APP_INITIALIZER` token](https://angular.io/api/core/APP_INITIALIZER) for executing initialization functions that complete before application initialization completes. However, in our case, we need the auth configuration _in order_ to initialize. So, we need to finish loading the configuration before initializing the app, which we can do when bootstrapping.
```

Comment:
This is how I do it in Ionic with Angular: 

APP_INITIALIZER: https://github.com/oktadev/schematics/blob/main/src/add-auth/ionic/angular/src/app/auth/auth.module.ts#L42

factory: https://github.com/oktadev/schematics/blob/main/src/add-auth/index.ts#L434